File: 225.用队列实现栈.js
/*
 * @lc app=leetcode.cn id=225 lang=javascript
 *
 * [225] 用队列实现栈
 */

// @lc code=start

var MyStack = function () {
  this.queue1 = [];
  this.queue2 = [];
  this.front1 = 0;
  this.front2 = 0;
  this.rear1 = 0;
  this.rear2 = 0;
};

/**
 * @param {number} x
 * @return {void}
 */
MyStack.prototype.push = function (x) {
  this.queue1[this.rear1++] = x;
};

/**
 * @return {number}
 */
MyStack.prototype.pop = function () {
  while (this.front1 + 1 !== this.rear1)
    this.queue2[this.rear2++] = this.queue1[this.front1++];
  const item = this.queue1[this.front1++];
  while (this.front2 !== this.rear2)
    this.queue1[this.rear1++] = this.queue2[this.front2++];
  return item;
};

/**
 * @return {number}
 */
MyStack.prototype.top = function () {
  return this.queue1[this.rear1 - 1];
};

/**
 * @return {boolean}
 */
MyStack.prototype.empty = function () {
  if (this.front1 === this.rear1 && this.front2 === this.rear2) return true;
  return false;
};

/**
 * Your MyStack object will be instantiated and called as such:
 * var obj = new MyStack()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.empty()
 */
// @lc code=end

File: 225.用队列实现栈.c
/*
 * @lc app=leetcode.cn id=225 lang=c
 *
 * [225] 用队列实现栈
 */

// @lc code=start

#define LEN 100

typedef struct
{
  int queue[LEN];
  int front;
  int rear;
} MyStack;

MyStack *myStackCreate()
{
  MyStack *stack = (MyStack *)malloc(sizeof(MyStack));
  stack->front = 0;
  stack->rear = 0;
  return stack;
}

void myStackPush(MyStack *obj, int x)
{
  if (obj->front != (obj->rear + 1) % LEN)
  {
    obj->queue[obj->rear] = x;
    obj->rear = (obj->rear + 1) % LEN;
  }
}

int myStackPop(MyStack *obj)
{
  int x, index;
  if (obj->front != obj->rear)
  {
    index = (obj->rear - 1) % LEN;
    while (obj->front != index)
    {
      myStackPush(obj, obj->queue[obj->front]);
      obj->front = (obj->front + 1) % LEN;
    }
    x = obj->queue[obj->front];
    obj->front = (obj->front + 1) % LEN;
    return x;
  }
  return NULL;
}

int myStackTop(MyStack *obj)
{
  int index;
  if (obj->front != obj->rear)
    return obj->queue[(obj->rear - 1) % LEN];
  return NULL;
}

bool myStackEmpty(MyStack *obj)
{
  if (obj->front == obj->rear)
    return true;
  return false;
}

void myStackFree(MyStack *obj)
{
  if (obj->front != obj->rear)
  {
    obj->front = 0;
    obj->rear = 0;
  }
}

/**
 * Your MyStack struct will be instantiated and called as such:
 * MyStack* obj = myStackCreate();
 * myStackPush(obj, x);

 * int param_2 = myStackPop(obj);

 * int param_3 = myStackTop(obj);

 * bool param_4 = myStackEmpty(obj);

 * myStackFree(obj);
*/
// @lc code=end
